middleware/openidconnect.go

```diff
@@ -0,0 +1,221 @@
+					w.WriteHeader(http.StatusInternalServerError)
+					return
+				}
+
+				opt.Logger.Debug().Str("body", string(body)).Msg("body")
+				switch strings.Split(res.Header.Get("Content-Type"), ";")[0] {
+				// application/jwt is in draft https://tools.ietf.org/html/draft-ietf-oauth-jwt-introspection-response-03
+				case "application/jwt":
+					// verify the jwt
+					// TODO this is a yet untested verification of jwt encoded introspection response
+
+					verifier := provider.Verifier(&oidc.Config{ClientID: opt.Audience})
+					idToken, err := verifier.Verify(customCtx, string(body))
+					if err != nil {
+						opt.Logger.Error().Err(err).Str("token", string(body)).Msg("could not verify jwt")
+						w.Header().Add("WWW-Authenticate", fmt.Sprintf(`Bearer realm="%s"`, opt.Realm))
+						http.Error(w, ErrInvalidToken.Error(), http.StatusUnauthorized)
+						return
+					}
+
+					if err := idToken.Claims(&claims); err != nil {
+						opt.Logger.Error().Err(err).Str("token", string(body)).Interface("id_token", idToken).Msg("failed to parse claims")
+						w.WriteHeader(http.StatusInternalServerError)
+						return
+					}
+				case "application/json":
+					var ir ocisoidc.IntrospectionResponse
+					// parse json
+					if err := json.Unmarshal(body, &ir); err != nil {
+						opt.Logger.Error().Err(err).Str("token", string(body)).Msg("failed to parse introspection response")
+						w.WriteHeader(http.StatusInternalServerError)
+						return
+					}
+					// verify the auth token is still active
+					if !ir.Active {
+						opt.Logger.Error().Interface("ir", ir).Str("body", string(body)).Msg("token no longer active")
+						w.Header().Add("WWW-Authenticate", fmt.Sprintf(`Bearer realm="%s"`, opt.Realm))
+						http.Error(w, ErrInvalidToken.Error(), http.StatusUnauthorized)
+						return
+					}
+					// resolve user info here? cache it?
+					oauth2Token := &oauth2.Token{
+						AccessToken: token,
+					}
+					userInfo, err := provider.UserInfo(customCtx, oauth2.StaticTokenSource(oauth2Token))
+					if err != nil {
+						opt.Logger.Error().Err(err).Str("token", string(body)).Msg("Failed to get userinfo")
+						w.WriteHeader(http.StatusInternalServerError)
+						return
+					}
+					if err := userInfo.Claims(&claims); err != nil {
+						opt.Logger.Error().Err(err).Interface("userinfo", userInfo).Msg("failed to unmarshal userinfo claims")
+						w.WriteHeader(http.StatusInternalServerError)
+						return
+					}
+					claims.Iss = ir.Iss
+					opt.Logger.Debug().Interface("claims", claims).Interface("userInfo", userInfo).Msg("unmarshalled userinfo")
+
+				default:
+					opt.Logger.Error().Str("content-type", res.Header.Get("Content-Type")).Msg("unknown content type")
+					w.WriteHeader(http.StatusInternalServerError)
+					return
+				}
+			}
+
+			// store claims in context
+			// uses the original context, not the one with probably reduced security
+			r.WithContext(ocisoidc.NewContext(r.Context(), &claims))
```

Comment:
WithContext creates a new request - which is not used - fishy ....

make staticcheck is bringing this up
```
✔ deepdiver:ocis-pkg [ oidc-middleware | ✔  ] ➭ make staticcheck 
go run honnef.co/go/tools/cmd/staticcheck -tags '' github.com/owncloud/ocis-pkg/log github.com/owncloud/ocis-pkg/middleware github.com/owncloud/ocis-pkg/oidc github.com/owncloud/ocis-pkg/service/debug github.com/owncloud/ocis-pkg/service/grpc github.com/owncloud/ocis-pkg/service/http github.com/owncloud/ocis-pkg/wrapper/prometheus
middleware/openidconnect.go:216:17: WithContext is a pure function but its return value is ignored (SA4017)
exit status 1
make: *** [Makefile:47: staticcheck] Error 1